tests/models/test_pyro.py

```diff
@@ -15,15 +16,51 @@
 from scvi.lightning import PyroTrainingPlan, Trainer
 
 
+class PyroJitGuideWarmup(Callback):
+    def __init__(self, train_dl) -> None:
+        super().__init__()
+        self.dl = train_dl
+
+    def on_train_start(self, trainer, pl_module):
+        """
+        Way to use PyroSample and have it be on GPU.
+
+        In this case, the PyroSample objects are added after the
+        model has already been moved to cuda.
+        """
+
+        # warmup guide for JIT
+        pyro_model = pl_module.module.model
+        dev = pyro_model.linear.weight.device
+        pyro_guide = pl_module.module.guide
+        for tensors in self.dl:
+            tens = {k: t.to(dev) for k, t in tensors.items()}
+            args, kwargs = pl_module.module._get_fn_args_from_batch(tens)
+            pyro_guide(*args, **kwargs)
+            break
+
+
 class BayesianRegressionPyroModel(PyroModule):
     def __init__(self, in_features, out_features):
         super().__init__()
+        self.in_features = in_features
+        self.out_features = out_features
 
         self.register_buffer("zero", torch.tensor(0.0))
         self.register_buffer("one", torch.tensor(1.0))
         self.register_buffer("ten", torch.tensor(10.0))
 
         self.linear = PyroModule[nn.Linear](in_features, out_features)
+        self.linear.weight = PyroSample(
+            lambda prior: dist.Normal(self.zero, self.one)
```

Comment:
This will work. Note it creates a reference cycle (self → linear → `._pyro_weight` → this lambda, whose closure → self), but you can probably ignore that.

Note the `prior` in `lambda prior` refers to `self.linear` (the owner of `.weight`). Our intended usage was to place any dynamic stuff on that owner, but I see how that would be cumbersome in your case, since the dynamic stuff lives in the root module.

Reply:
> Note the prior in lambda prior refers to self.linear (the owner of .weight). Our intended usage was to place any dynamic stuff on that owner, but I see how that would be cumbersome in your case, since the dynamic stuff lives in the root module.

I had noticed that... happy that this works though. Is it more desirable than the callback I had shown you earlier?

Reply:
Yes, your usage in this PR is pretty clean, I believe recommended over the callback you showed me.

Reply:
Thanks!